libraries/lib-string-utils/FromChars.cpp

```diff
@@ -0,0 +1,190 @@
+/*!********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file FromChars.cpp
+ @brief Define functions to convert numeric types to string representation.
+
+ Dmitry Vedenko
+ **********************************************************************/
+
+#include "FromChars.h"
+
+#include "3party/fast_float.h"
+
+#include <limits>
+#include <algorithm>
+
+namespace
+{
+// Converting c to unsigned here helps to eliminate one check 
+unsigned digitToInt(char c) noexcept
+{
+   return unsigned(c - '0');
+}
+
+template <typename ResultType>
+FromCharsResult FastStringToInt(
+   const char* first, const char* last, ResultType& value,
+   bool isNegative) noexcept
+{
+   using UnsignedResultType = std::make_unsigned_t<ResultType>;
+
+   const auto availableBytes = last - first;
+
+   if (availableBytes <= 0)
+      return { first, std::errc::invalid_argument };
+
+   UnsignedResultType result = digitToInt(*first);
+
+   if (result > 10)
+      return { first, std::errc::invalid_argument };
+
+   constexpr auto maxSafeDigits = std::numeric_limits<ResultType>::digits10 - 1;
+
+   const char* ptr = first;
+   const char* safeLast =
+      first + std::min<decltype(availableBytes)>(availableBytes, maxSafeDigits);
+
+   unsigned d;
+
+   // No integer overflow here
+   while (++ptr < safeLast && (d = digitToInt(*ptr)) <= 9)
+   {
+      result = result * 10 + d;
+   }
+
+   // But here live dragons
+   while (ptr < last && (d = digitToInt(*ptr++)) <= 9)
+   {
+      const UnsignedResultType maxValue =
+         std::is_signed_v<ResultType> && isNegative ?
+            static_cast<UnsignedResultType>(
+               std::numeric_limits<ResultType>::max() + 1) :
+            std::numeric_limits<ResultType>::max();
+
+      const auto temp = result * 10 + d;
+
+      if (temp < result || temp > maxValue) // Overflow!
```

Comment:
temp > maxValue is useless in the unsigned case.

temp < result is an insufficient check for overflow!

Try it for example with ResultType as unsigned char, and the string "724".  You will get 212 and no error detected.

Please review another suggested commit from me to fix this.

Reply:
> temp < result is an insufficient check for overflow!

Hmm, it will, but probably not for the `uint64_t`, let me think. Definitely works for  `uint8_t`:

https://crsib-screenshots.s3.amazonaws.com/202111161744-bs5nq.png

>temp > maxValue is useless in the unsigned case.

We can compile it out by if constexpr, but I don't think if it really matters though